FR #976: Add support for contract commitments (#1350)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/contract_commitment/columns/contractcommitmentid.md

# Contract Commitment ID

> Contract Commitment ID is a provider-assigned identifier describing an agreement negotiated between a provider and a customer.  Contracts can include commitment to a certain amount of spend or usage over an agreed period of time.

Reviewer: Contract application properties within Contract Applied are defined as follows:

> * `Contract ID`: The unique identifier representing a single contract.
> * `Contract Commitment ID`: The unique identifier representing a single contract term.

`Contract ID` is defined as:

> Contract ID is a provider-assigned identifier for a contract describing the agreed terms between a provider and a customer.

Following the same pattern, I suggest we consider the following formulation for `Contract Commitment ID`, primarily to avoid using the term *negotiated*, as we've previously agreed not to constrain this iteration in that way:

```suggestion
Contract Commitment ID is a provider-assigned identifier describing a single contract term agreed between a provider and a customer.  Contracts can include commitments to a certain amount of spend or usage over an agreed period of time.
```

Author: Definitely!  Thanks for this catch.  I thought I had scrubbed use of the term "negotiated", given that we will support contract commitments that aren't negotiated.  Committing!

> The ContractCommitmentID column adheres to the following requirements:

Reviewer 2: What's the relevance of 'commitment' here? Is this not relevant if there's no particular spend commitment?

I would have thought it would be valuable to simply point to the contract governing the usage. Then, you're able to use this to analyze cost regardless of if there's a commitment or not.

Reviewer 2: maybe this is a duplicate? I see a contract ID below with the same description.

Reviewer 2: Seems like this might be too early to comment :) I'll leave this here and will check back in later.

Author: @UdamDewaraja This entity is meant to represent one of many agreements on the contract; a contract line item, if you will.  It's a child of Contract ID.  We can certainly discuss alternate names if `commitment` isn't your preference.  I can see making that change if it's just a benefit you get as a result of commitments on the contract.  The joke example @mike-finopsorg and I were using was "get a free box of cigars every year".

The idea here is to track both the overall contract value as a row in this dataset (where Contract ID = Contract Commitment ID), as well as the various other line items on that contract that might represent separate commitments, and do not necessarily add up to that overall value.  Think EDP vs an S3 commitment as part of that EDP.   If we don't do it this way, then when we model that S3 commitment later, the practitioner would have to look in two different places to see all contract commitments.  This puts it in one place.

I leaned on TF1's feedback from last week, as well as a session with @mike-finopsorg, to arrive at this design.  Would love your attendance in TF1 this week and/or async feedback here!

Author: Going to resolve this comment -- feel free to unresolve if you want to continue this thread!

* ContractCommitmentID MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* ContractCommitmentID MUST be of type String.

> * ContractCommitmentID MUST conform to [StringHandling](#stringhandling) requirements.

Reviewer: ```suggestion
The ContractCommitmentId column adheres to the following requirements:

* ContractCommitmentId MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* ContractCommitmentId MUST be of type String.
* ContractCommitmentId MUST conform to [StringHandling](#stringhandling) requirements.
```

> * ContractCommitmentID nullability is defined as follows:
>   * ContractCommitmentID MUST be null when a [*charge*](#glossary:charge) is not related to a *contract commitment*.
>   * ContractCommitmentID MUST NOT be null when a *charge* is related to a *contract commitment*.

Reviewer: ```suggestion
* ContractCommitmentId MUST NOT be null.
```

* When ContractCommitmentID is not null, ContractCommitmentID adheres to the following additional requirements:
  * ContractCommitmentID MUST be a unique identifier within the provider.
  * ContractCommitmentID SHOULD be a fully-qualified identifier.
* ContractCommitmentID MUST have one and only one parent [ContractID](#contractid).
* ContractCommitmentID MAY be equal to ContractID.

> * ContractCommitmentID MUST be unique across the Contract Commitment dataset.

Reviewer: ```suggestion
* When ContractCommitmentId is not null, ContractCommitmentId adheres to the following additional requirements:
  * ContractCommitmentId MUST be a unique identifier within the provider.
  * ContractCommitmentId SHOULD be a fully-qualified identifier.
* ContractCommitmentId MUST have one and only one parent [ContractId](#contractid).
* ContractCommitmentId MAY be equal to ContractId.
* ContractCommitmentId MUST be unique across the Contract Commitment dataset.
```

## Column ID

> ContractCommitmentID

Reviewer: ```suggestion
ContractCommitmentId
```

## Display Name

Contract Commitment ID

## Description

> A provider-assigned identifier describing an agreement negotiated between a provider and a customer.

Reviewer: see [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1350/files#r2349846854)

Author: Agreed.  Let's go with:

```suggestion
A provider-assigned identifier describing a single contract term agreed between a provider and a customer.
```
